vm/systemSmartContracts/governance.go

```diff
@@ -288,521 +499,599 @@ func (g *governanceContract) whiteListProposal(args *vmcommon.ContractCallInput)
+	if len(args.Arguments) != 0 {
+		log.Warn("excepted argument number is 0")
 		return vmcommon.UserError
 	}
-
-	numNodesToVote := int32(0)
-	validatorData, err := g.getOrCreateValidatorData(validatorAddress, int32(numStakedNodes))
-	if err != nil {
-		log.Warn("getOrCreateValidatorData", "err", err)
-		g.eei.AddReturnMessage("getOrCreateValidator data error" + err.Error())
+	if g.proposalExists(args.CallerAddr) {
+		log.Warn("proposal with this key already exists")
 		return vmcommon.UserError
 	}
 
-	found := false
-	for _, voter := range validatorData.Delegators {
-		if bytes.Equal(voter.Address, voterAddress) {
-			found = true
-			numNodesToVote = voter.NumNodes
-			break
-		}
+	key := append([]byte(proposalPrefix), args.CallerAddr...)
+	whiteListAcc := &WhiteListProposal{
+		WhiteListAddress: args.CallerAddr,
+		ProposalStatus:   key,
 	}
-	if !found || numNodesToVote <= 0 {
-		g.eei.AddReturnMessage("address has 0 voting power")
+
+	key = append([]byte(whiteListPrefix), args.CallerAddr...)
+	generalProposal := &GeneralProposal{
+		IssuerAddress:  args.CallerAddr,
+		GitHubCommit:   []byte("genesis"),
+		StartVoteNonce: 0,
+		EndVoteNonce:   0,
+		Yes:            g.governanceConfig.MinQuorum,
+		No:             big.NewInt(0),
+		Veto:           big.NewInt(0),
+		Voted:          true,
+		Votes:          make([][]byte, 0),
+	}
+	marshaledData, err := g.marshalizer.Marshal(whiteListAcc)
+	if err != nil {
+		log.Warn("marshal error in whiteListAtGenesis", "err", err)
 		return vmcommon.UserError
 	}
+	g.eei.SetStorage(key, marshaledData)
 
-	err = g.voteForProposal(proposalToVote, voteString, voterAddress, numNodesToVote)
+	err = g.saveGeneralProposal(args.CallerAddr, generalProposal)
 	if err != nil {
-		g.eei.AddReturnMessage("voteForProposal " + err.Error())
+		log.Warn("save general proposal ", "err", err)
 		return vmcommon.UserError
 	}
 
 	return vmcommon.Ok
 }
 
-func (g *governanceContract) isValidVoteString(vote string) bool {
-	switch vote {
-	case "yes":
-		return true
-	case "no":
-		return true
-	case "veto":
-		return true
-	case "dontCare":
-		return true
-	}
-	return false
+// applyVote takes in a vote and a full VoteData object and correctly applies the new vote, then returning
+//  the new full VoteData object. In the same way applies the vote to the general proposal
+func (g *governanceContract) applyVote(vote *VoteDetails, voteData *VoteData, proposal *GeneralProposal) (*VoteData, *GeneralProposal, error) {
+	switch vote.Value {
+	case Yes:
+		voteData.TotalYes.Add(voteData.TotalYes, vote.Power)
+		proposal.Yes.Add(proposal.Yes, vote.Power)
+		break
+	case No:
+		voteData.TotalNo.Add(voteData.TotalNo, vote.Power)
+		proposal.No.Add(proposal.No, vote.Power)
+		break
+	case Veto:
+		voteData.TotalVeto.Add(voteData.TotalVeto, vote.Power)
+		proposal.Veto.Add(proposal.Veto, vote.Power)
+		break
+	default:
+		return nil, nil, fmt.Errorf("%s: %s", vm.ErrInvalidArgument, "invalid vote type")
```

Comment:
I know, but I think it's better to have it rather than not. In the case of refactoring over time we might not need those checks before the function call and end up with bugs.

Reply:
ok